packages/core/lib/segments/plugins/plugin.js

```diff
@@ -17,30 +24,75 @@ var Plugin = {
         res.on('data', function(chunk) {
           body += chunk;
         });
+
         res.on('end', function() {
           if (this.statusCode === 200 || this.statusCode === 300) {
-            body = JSON.parse(body);
+            try {
+              body = JSON.parse(body);
+            } catch (e) {
+              callback(e);
+              return;
+            }
+            
             callback(null, body);
-          } else if (retries > 0 && this.statusCode === 400){
+          } else if (retries > 0 && Math.floor(this.statusCode / 100) === 5) {
             retries--;
             setTimeout(getMetadata, METADATA_RETRY_TIMEOUT);
-          } else { callback(); }
+          } else {
+            callback(new Error(`Failed to retrieve metadata with options: ${options}`));
+          }
         });
-      }).on('error', function(err) {
+      });
+
+      req.on('error', function(err) {
         callback(err);
       });
 
-      req.on('socket', function(socket) {
-        socket.on('timeout', function() {
-          req.abort();
-        });
-        socket.setTimeout(METADATA_TIMEOUT);
+      req.on('timeout', function() {
+        req.abort();
       });
 
+      req.setTimeout(Plugin.METADATA_TIMEOUT);
       req.end();
     };
 
     getMetadata();
+  },
+
+  /**
+   * Asynchronously retrieves a token used in requests to EC2 instance metadata service.
```

Comment:
I think this method is passing `option` to the request but doesn't really have any logic specific to retrieving the token. How about removing the `options` callback and setting up the request here? The options look like constants that can be inlined here and then this method does become a `getToken` method :)

Reply:
I agree with anuraag here that this function doesn't really do anything specific to token. Also, token is currently only used with IMDSv2 which is only available for EC2, so if you're considering to make this function into a real `getToken` function, maybe moving it to `ec2_plugin.js` would make more sense.

Reply:
Both fair callouts, I'll move it to `ec2_plugin` and narrow its scope.